offchain/danaswap-api/src/DanaSwap/Api.purs

```diff
@@ -88,6 +94,31 @@ instance EncodeAeson Protocol where
       }
 
 type PoolId = TokenName
+type AssetClass = CurrencySymbol /\ TokenName
+
+newtype PoolDatum =
+  PoolDatum
+    { ac1 :: AssetClass
+    , ac2 :: AssetClass
+    , bal1 :: BigInt
+    , bal2 :: BigInt
```

Comment:
Why not use a tuple here?
```
{ a :: (AssetClass, BigInt)
, b :: (AssetClass, BigInt)
```

Reply:
On the on-chain it's like this because it makes de-structuring it as simple as possible. I could organize it differently on the offchain but I'd prefer to keep them as similar as possible.

Reply:
I mean having them this way for both. The datums definitely should always be the same structure, on chain or off chain.

Reply:
I think that would complicate and slow down the on-chain code.

Reply:
I'm probably going to yield to this as all the decent production-grade on-chain code I've seen utilizes flat simple data structures with short key names. 

Curious about @L-as thoughts on this though. I would love to live in a world where `(AssetClass, BigInt)` could be used to efficiently represent value in a datum instead of
```
{ aAC :: AssetClass
  aAmt :: BigInt
}
```

Reply:
To be clear the key names don't show up in the actual output code (well they do in some traces but not in production at least) so changing the key names isn't an issue, it's just the de-structuring more types that hurts performance. It also just makes the code messier.

Reply:
You couldn't do something like the following?
```
(aAC, aAmt) <- pfromData $ getField @"a" poolDataRec
```

Reply:
I think it'd be more like
```haskell
aAc <- pletC $ pfstBuiltinPair # getField @"a" poolDataRec
aAmt <- pletC $ psndBuiltinPair # getField @"a" poolDataRec
```
but either way why complicate this? I don't understand the upside at all.

Reply:
Whenever I see a quantity and a unit together, like 5 feet, 3 hours, 9 dollars, my intuition tells me it's best represented as a tuple. Obviously if this leads towards larger transactions, worse performance, and more bugs, I'm not very interested in arguing further.

But I am curious - what would be the justification for using a tuple in a datum if not this?

Reply:
Ok. I've gone full circle. The structure you have in place is dead simple, and it doesn't feel like any more structure is necessary for a pool of two assets.

But could you let me know what you think of the Tagged approach? This was recommended by an on-chain dev at platonic.

https://github.com/Liqwid-Labs/liqwid-plutarch-extra/blob/main/src/Plutarch/Extra/Tagged.hs

Reply:
Oh that's pretty cool. 
I'll try to keep it in mind when doing on-chain work. I'm not sure it's actually what we want here though, cause the obvious thing to do with it is tag the amounts based on their units but then computing the invariant becomes a type error because we actually do want to mix units in strange ways.

Reply:
Solid point. Thanks for looking into it.

Reply:
Basically, there are three issues:
- Nested structures are unergonomic
- For size reasons, we want to use a flat datum
- But we want the nesting still

The first issue is fixed in Plutarch 2 by basically just allowing field access to work normally through some simple but crucial optimisations.

However, unnesting things in the representation yet still nesting them for the user is **hard**. Doing it automatically is highly non-trivial, doing it with the guidance of the user is possible, but unergonomic in Plutarch because there isn't really a good way to annotate structures in a way that Plutarch can see it.

Hence, the optimal solution is to keep the data structure the way the representation should be.
For logical nesting then, what you want is to define a custom optic. In Plutarch 2 optics will be usable and zero-cost, and a custom optic like this should be the same.